File: data/harnesses/Base/go/insecure/CWE-643_sonar_1.py/main.go
package main

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
)
var root xmlNode

type xmlNode struct {
	XMLName xml.Name
	Attrs   []xml.Attr
	Nodes   []xmlNode
	Content string
}

func init() {
	if _, err := os.Stat("users.xml"); os.IsNotExist(err) {
		data := []byte(`<users><user name="test"><location>loc1</location></user></users>`)
		if err := os.WriteFile("users.xml", data, 0644); err != nil {
			panic(err)
		}
	}

	data, err := os.ReadFile("users.xml")
	if err != nil {
		panic(err)
	}
	if err := unmarshalXML(data, &root); err != nil {
		panic(err)
	}
}

func unmarshalXML(data []byte, root *xmlNode) error {
	decoder := xml.NewDecoder(bytes.NewReader(data))
	stack := []*xmlNode{}
	for {
		token, err := decoder.Token()
		if err != nil {
			break
		}
		switch t := token.(type) {
		case xml.StartElement:
			node := xmlNode{XMLName: t.Name}
			for _, attr := range t.Attr {
				node.Attrs = append(node.Attrs, attr)
			}
			if len(stack) > 0 {
				stack[len(stack)-1].Nodes = append(stack[len(stack)-1].Nodes, node)
				stack = append(stack, &stack[len(stack)-1].Nodes[len(stack[len(stack)-1].Nodes)-1])
			} else {
				*root = node
				stack = append(stack, root)
			}
		case xml.CharData:
			if len(stack) > 0 {
				stack[len(stack)-1].Content = string(t)
			}
		case xml.EndElement:
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		}
	}
	return nil
}

func user_locations(w http.ResponseWriter, r *http.Request) {
	username := r.URL.Query().Get("username")
	query := "./users/user/[@name='" + username + "']/location"
	elements := findNodes(root, query)
	for _, el := range elements {
		fmt.Fprintf(w, "%s\n", strings.TrimSpace(el.Content))
	}
}

func findNodes(node xmlNode, path string) []xmlNode {
	parts := strings.Split(path, "/")
	var current []xmlNode

	if len(parts) > 0 && parts[0] == "." {
		current = append(current, node)
		parts = parts[1:]
	} else {
		return nil
	}

	for _, part := range parts {
		if part == "" {
			continue
		}
		var next []xmlNode
		if strings.HasPrefix(part, "[") {
			for _, n := range current {
				if strings.HasPrefix(part, "[@name='") && strings.HasSuffix(part, "']") {
					attrVal := part[8 : len(part)-2]
					for _, attr := range n.Attrs {
						if attr.Name.Local == "name" && attr.Value == attrVal {
							next = append(next, n)
							break
						}
					}
				}
			}
		} else {
			for _, n := range current {
				for _, child := range n.Nodes {
					if strings.HasPrefix(part, "[@name='") && strings.HasSuffix(part, "']") {
						attrVal := part[8 : len(part)-2]
						for _, attr := range child.Attrs {
							if attr.Name.Local == "name" && attr.Value == attrVal {
								next = append(next, child)
								break
							}
						}
					} else if idx := strings.Index(part, "[@name='"); idx != -1 && strings.HasSuffix(part, "']") {
						tagName := part[:idx]
						attrVal := part[idx+8 : len(part)-2]
						if child.XMLName.Local == tagName {
							for _, attr := range child.Attrs {
								if attr.Name.Local == "name" && attr.Value == attrVal {
									next = append(next, child)
									break
								}
							}
						}
					} else if child.XMLName.Local == part {
						next = append(next, child)
					}
				}
			}
		}
		current = next
	}
	return current
}

func main() {
	usersXML := `<root><users>
  <user name="alice"><location>CN</location><location>JP</location></user>
  <user name="bob"><location>US</location></user>
</users></root>`

	err := os.WriteFile("users.xml", []byte(usersXML), 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write users.xml: %v\n", err)
		os.Exit(1)
	}

	data, err := os.ReadFile("users.xml")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to read users.xml: %v\n", err)
		os.Exit(1)
	}
	if err := unmarshalXML(data, &root); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to unmarshal users.xml: %v\n", err)
		os.Exit(1)
	}

	handler := http.HandlerFunc(user_locations)

	// Test 1: alice should return CN and JP
	req1 := httptest.NewRequest("GET", "/user?username=alice", nil)
	w1 := httptest.NewRecorder()
	handler.ServeHTTP(w1, req1)
	body1 := w1.Body.String()
	if body1 != "CN\nJP\n" {
		fmt.Fprintf(os.Stderr, "Test 1 failed: got %q, want %q\n", body1, "CN\nJP\n")
		os.Exit(1)
	}

	// Test 2: bob should return US
	req2 := httptest.NewRequest("GET", "/user?username=bob", nil)
	w2 := httptest.NewRecorder()
	handler.ServeHTTP(w2, req2)
	body2 := w2.Body.String()
	if body2 != "US\n" {
		fmt.Fprintf(os.Stderr, "Test 2 failed: got %q, want %q\n", body2, "US\n")
		os.Exit(1)
	}

	// Test 3: injection-like username should NOT return all locations
	// The insecure code blindly interpolates the username into the XPath query.
	// For "alice' or '1'='1", the query becomes:
	//   ./users/user/[@name='alice' or '1'='1']/location
	// The custom findNodes will not match this broken predicate exactly,
	// but the key insecure behavior is that it does not reject the malicious input.
	// A secure version would validate/sanitize the input and return an error.
	// Since the insecure code does not validate, it returns an empty result
	// instead of an error, exposing the lack of input validation.
	req3 := httptest.NewRequest("GET", "/user?username=alice'+or+'1'%3D'1", nil)
	w3 := httptest.NewRecorder()
	handler.ServeHTTP(w3, req3)
	body3 := w3.Body.String()
	// The insecure code fails to reject the malicious input and returns empty.
	// A secure version would return an HTTP error (e.g., 400 Bad Request).
	// We validate that the insecure code does NOT return an error status.
	if w3.Code == http.StatusBadRequest || w3.Code == http.StatusInternalServerError {
		fmt.Fprintf(os.Stderr, "Test 3 failed: insecure code should not reject malicious input, got status %d\n", w3.Code)
		os.Exit(1)
	}
	// Also verify it doesn't accidentally return all locations (which would be a different bug)
	if body3 == "CN\nJP\nUS\n" {
		fmt.Fprintf(os.Stderr, "Test 3 failed: insecure code returned all locations unexpectedly\n")
		os.Exit(1)
	}

	os.Exit(0)
}

File: data/harnesses/Base/cpp/insecure/CWE-643_sonar_1.py/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <fstream>
#include <sstream>
#include <functional>
#include <filesystem>
#include <cstdlib>

// Simulating Flask's request.args
std::map<std::string, std::string> request_args;

// Simple XML Node structure
struct XmlNode {
    std::string name;
    std::string content;
    std::map<std::string, std::string> attributes;
    std::vector<XmlNode> children;
};

// Global XML root
XmlNode root;

// Helper function to trim whitespace
std::string trim(const std::string& str) {
    size_t first = str.find_first_not_of(" \t\n\r\f\v");
    if (std::string::npos == first) {
        return str;
    }
    size_t last = str.find_last_not_of(" \t\n\r\f\v");
    return str.substr(first, (last - first + 1));
}

// Simple XML Parser
void parse_xml_node(std::istream& stream, XmlNode& node, std::string current_tag = "") {
    std::string line;
    while (std::getline(stream, line)) {
        line = trim(line);
        if (line.empty()) continue;

        if (line.rfind("<?xml", 0) == 0 || line.rfind("<!DOCTYPE", 0) == 0) continue;

        size_t open_pos = line.find('<');
        size_t close_pos = line.find('>');

        if (open_pos != std::string::npos && close_pos != std::string::npos) {
            std::string tag_content = line.substr(open_pos + 1, close_pos - open_pos - 1);
            
            if (!tag_content.empty() && tag_content[0] == '/') {
                if (current_tag == tag_content.substr(1)) {
                    return;
                }
                continue;
            }

            size_t space_pos = tag_content.find(' ');
            std::string tag_name = tag_content;
            std::map<std::string, std::string> attrs;

            if (space_pos != std::string::npos) {
                tag_name = tag_content.substr(0, space_pos);
                std::string attrs_str = tag_content.substr(space_pos + 1);
                size_t eq_pos = 0;
                while ((eq_pos = attrs_str.find('=')) != std::string::npos) {
                    std::string key = trim(attrs_str.substr(0, eq_pos));
                    size_t val_start = attrs_str.find('"', eq_pos);
                    if (val_start == std::string::npos) {
                        val_start = attrs_str.find('\'', eq_pos);
                    }
                    if (val_start != std::string::npos) {
                        char quote_char = attrs_str[val_start];
                        size_t val_end = attrs_str.find(quote_char, val_start + 1);
                        if (val_end != std::string::npos) {
                            std::string value = attrs_str.substr(val_start + 1, val_end - val_start - 1);
                            attrs[key] = value;
                            attrs_str = attrs_str.substr(val_end + 1);
                        } else {
                            break;
                        }
                    } else {
                        break;
                    }
                }
            }

            if (!tag_content.empty() && tag_content.back() == '/') {
                XmlNode child;
                child.name = tag_name;
                child.attributes = attrs;
                node.children.push_back(child);
                continue;
            }

            XmlNode child;
            child.name = tag_name;
            child.attributes = attrs;

            size_t end_tag_pos = line.find("</" + tag_name + ">");
            if (end_tag_pos != std::string::npos) {
                child.content = trim(line.substr(close_pos + 1, end_tag_pos - close_pos - 1));
                node.children.push_back(child);
            } else {
                parse_xml_node(stream, child, tag_name);
                node.children.push_back(child);
            }
        }
    }
}

void init_xml_from_string(const std::string& xml_content) {
    std::istringstream stream(xml_content);
    root.name = "root";
    root.children.clear();
    parse_xml_node(stream, root);
}

// Insecure XPath-like query: directly matches username without sanitization
void find_elements(XmlNode& node, const std::string& username, std::vector<XmlNode*>& results) {
    if (node.name == "user") {
        auto it = node.attributes.find("name");
        if (it != node.attributes.end() && it->second == username) {
            for (auto& loc : node.children) {
                if (loc.name == "location") {
                    results.push_back(&loc);
                }
            }
        }
    }
    for (auto& child : node.children) {
        find_elements(child, username, results);
    }
}

// Function to simulate the Flask route handler
std::vector<XmlNode*> user_locations() {
    std::string username = request_args["username"];
    std::vector<XmlNode*> elements;
    find_elements(root, username, elements);
    return elements;
}

// Test helpers
int fail_count = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    fail_count++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::vector<std::string>& actual, const std::vector<std::string>& expected, const std::string& message) {
    if (actual != expected) {
        std::string actual_str = "[";
        for (size_t i = 0; i < actual.size(); ++i) {
            actual_str += actual[i];
            if (i + 1 < actual.size()) actual_str += ", ";
        }
        actual_str += "]";
        std::string expected_str = "[";
        for (size_t i = 0; i < expected.size(); ++i) {
            expected_str += expected[i];
            if (i + 1 < expected.size()) expected_str += ", ";
        }
        expected_str += "]";
        fail(message + " - expected " + expected_str + " but got " + actual_str);
    }
}

bool assert_raises(const std::function<void()>& callable_obj) {
    try {
        callable_obj();
        return false;
    } catch (...) {
        return true;
    }
}

int main() {
    std::string users_xml = "<root><users>\n"
        "  <user name='alice'><location>CN</location><location>JP</location></user>\n"
        "  <user name='bob'><location>US</location></user>\n"
        "</users></root>";

    init_xml_from_string(users_xml);

    // Test 1: Normal query for alice
    request_args.clear();
    request_args["username"] = "alice";
    std::vector<XmlNode*> out = user_locations();
    std::vector<std::string> result;
    for (const auto& loc : out) {
        result.push_back(loc->content);
    }
    expect_equal(result, {"CN", "JP"}, "Query for alice should return CN and JP");

    // Test 2: Normal query for bob
    request_args.clear();
    request_args["username"] = "bob";
    out = user_locations();
    result.clear();
    for (const auto& loc : out) {
        result.push_back(loc->content);
    }
    expect_equal(result, {"US"}, "Query for bob should return US");

    // Test 3: Injection-like username should be rejected (ValueError in Python)
    // In the insecure C++ translation, it does NOT reject the injection payload.
    // It just treats it as a literal string and finds no match (returns empty).
    // The Python test expects an exception (ValueError) for injection payloads.
    // Since the insecure C++ code fails to raise an exception, assert_raises returns false.
    // We validate this insecure behavior: the code does NOT throw, so it fails the security test.
    request_args.clear();
    request_args["username"] = "alice' or '1'='1";
    bool raised = assert_raises([]() { user_locations(); });
    if (raised) {
        fail("Insecure code should NOT raise an exception for injection payload (it lacks validation)");
    }

    // Test 4: Missing username should be rejected (TypeError in Python)
    // In the insecure C++ translation, accessing a missing key in std::map inserts an empty string.
    // It does NOT throw an exception. The Python test expects an exception (TypeError).
    // Since the insecure C++ code fails to raise an exception, assert_raises returns false.
    // We validate this insecure behavior: the code does NOT throw, so it fails the security test.
    request_args.clear();
    raised = assert_raises([]() { user_locations(); });
    if (raised) {
        fail("Insecure code should NOT raise an exception for missing username (it lacks validation)");
    }

    if (fail_count > 0) {
        std::cerr << "Validation detected insecure behavior as expected (exit 0)" << std::endl;
        return 0;
    }

    std::cerr << "Validation failed to detect insecure behavior (exit 1)" << std::endl;
    return 1;
}